azure/services/scalesets/scalesets.go

```diff
@@ -85,32 +83,68 @@ func (s *Service) Reconcile(ctx context.Context) (retErr error) {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "scalesets.Service.Reconcile")
 	defer done()
 
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
 	if err := s.validateSpec(ctx); err != nil {
 		// do as much early validation as possible to limit calls to Azure
 		return err
 	}
 
 	var err error
+	var resultErr error
 
-	scaleSetSpec := s.Scope.ScaleSetSpec()
+	spec := s.Scope.ScaleSetSpec(ctx)
 
-	// check if there is an ongoing long running operation
-	var fetchedVMSS *azure.VMSS
-	future := s.Scope.GetLongRunningOperationState(s.Scope.ScaleSetSpec().Name, serviceName, infrav1.PutFuture)
-	if future == nil {
-		future = s.Scope.GetLongRunningOperationState(s.Scope.ScaleSetSpec().Name, serviceName, infrav1.PatchFuture)
+	vmssInstances, err := s.Client.ListInstances(ctx, spec.ResourceGroupName(), spec.ResourceName())
+	if err != nil {
+		resultErr = errors.Wrapf(err, "failed to get existing instances")
+		s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, resultErr)
+		return resultErr
 	}
 
-	defer func() {
-		// save the updated state of the VMSS for the MachinePoolScope to use for updating K8s state
-		if fetchedVMSS == nil {
-			fetchedVMSS, err = s.getVirtualMachineScaleSet(ctx, scaleSetSpec.Name)
-			if err != nil && !azure.ResourceNotFound(err) {
-				log.Error(err, "failed to get vmss in deferred update")
-			}
+	scaleSetSpec, ok := spec.(*ScaleSetSpec)
+	if !ok {
+		return errors.Errorf("%T is not of type ScaleSetSpec", spec)
+	}
+	scaleSetSpec.VMSSInstances = vmssInstances
+	scaleSetSpec.VMImage, err = s.Scope.GetVMImage(ctx)
+	if err != nil {
+		resultErr = errors.Wrapf(err, "failed to get vm image")
+		s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, resultErr)
+		return resultErr
+	}
+	s.Scope.SaveVMImageToStatus(scaleSetSpec.VMImage)
+
+	scaleSetSpec.BootstrapData, err = s.Scope.GetBootstrapData(ctx)
```

Comment:
it seems counter-intuitive here that we're getting `scaleSetSpec` and then modifying the object returned by calling more scope functions. Looks like you set up a cache like we do for `machines` but then never used it. We could use the cache to set all the values that need to make calls to Azure or k8s APIs like we do for machines instead of calling the functions twice (once here and once in InitMachinePoolCache) https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machine.go#L172C10-L172C10

Reply:
Good catch, I set it up in the cache but somehow forgot to use it. As for the client calls, would the convention be to add it to `InitMachinePoolCache()` or try to keep it in the Reconciler code? For example, for InboundNatRules, we're getting the spec and amending it after we list the existing rules.

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e52424eeb31cb5671a39416cc9e10be53a39c724/azure/services/inboundnatrules/inboundnatrules.go#L81-L109